### PR TITLE
run happy path tests on push; bump hermes version

### PR DESCRIPTION
**Pull request by MSalopek**
# Description

This PR bumps hermes version to v1.2.0 and runs only happy path integration tests during pipeline execution.

After each testrun its corresponding docker container is removed to avoid running unused chains in the background.

Subsequent PRs will allow running the full integration test suite as a cronjob.

## Linked issues

Part 1/2 for closing: #631 

## Type of change

If you've checked more than one of the first three boxes, consider splitting this PR into multiple PRs!

- [ ] `Feature`: Changes and/or adds code behavior, irrelevant to bug fixes
- [ ] `Fix`: Changes and/or adds code behavior, specifically to fix a bug
- [x] `Refactor`: Changes existing code style, naming, structure, etc.
- [ ] `Testing`: Adds testing
- [ ] `Docs`: Adds documentation

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -22,19 +22,13 @@ RUN go mod tidy
...
+FROM informalsystems/hermes:1.2.0 AS hermes-builder
 
 FROM --platform=linux/amd64 fedora:36
 RUN dnf update -y
-RUN dnf install -y which iproute iputils procps-ng vim-minimal tmux net-tools htop jq openssl1.1
+RUN dnf install -y which iproute iputils procps-ng vim-minimal tmux net-tools htop jq
 USER root
 
-# Copy Hermes and IS binaries to final image
-COPY --chown=0:0 --from=hermes-builder /usr/lib/x86_64-linux-gnu/libssl.so.1.1 /usr/lib/x86_64-linux-gnu/libssl.so.1.1
```
Comment (MSalopek) [inline Dockerfile:35]: This dependency is no longer required. Hermes seems to have either bundled libssl with the binary or no longer depends on this specific version

File: Makefile [hunk 2]
```diff
@@ -22,7 +22,11 @@ test-diff:
 
 # run integration tests
 test-integration:
```
Comment (smarshall-spitzbart) [inline Makefile:24]: Might be better to make this `test-integration-short` and the other, `test-integration` to stay consistent with the pattern used by `test` and `test-short`

(would require a change in yml file too)

File: tests/integration/main.go [hunk 3]
```diff
@@ -14,7 +14,8 @@ import (
 )
 
 var verbose = flag.Bool("verbose", false, "turn verbose logging on/off")
-var multiconsumer = flag.Bool("multiconsumer", false, "run happy path and multiconsumer tests in parallel")
+var happyPathOnly = flag.Bool("happy-path-only", false, "run happy path tests only")
+var multiconsumer = flag.Bool("multiconsumer", false, "include multiconsumer tests in run")
```
Comment (smarshall-spitzbart) [inline tests/integration/main.go:18]: Flag naming consistency here could be nice 
```suggestion
var multi-consumer = flag.Bool("multiconsumer", false, "include multiconsumer tests in run")
```

File: tests/integration/main.go [hunk 4]
```diff
@@ -24,45 +25,47 @@ var localSdkPath = flag.String("local-sdk-path", "",
 func main() {
 	flag.Parse()
 
-	// wg waits for all runners to complete
-	var wg sync.WaitGroup
+	if happyPathOnly != nil && *happyPathOnly {
+		tr := DefaultTestRun()
```
Comment (smarshall-spitzbart) [inline tests/integration/main.go:29]: Is there a short log saying only happy path tests are running?

File: tests/integration/main.go [hunk 5]
```diff
@@ -131,9 +134,14 @@ func (tr *TestRun) runStep(step Step, verbose bool) {
 	}
 }
 
-// Starts docker container and sequentially runs steps
-func (tr *TestRun) ExecuteSteps(wg *sync.WaitGroup, steps []Step) {
+// ExecuteStepsInWaitGroup runs steps in a WaitGroup so that the test run can be run in parallel.
+func (tr *TestRun) ExecuteStepsInWaitGroup(wg *sync.WaitGroup, steps []Step) {
 	defer wg.Done()
+	tr.ExecuteSteps(steps)
```
Comment (smarshall-spitzbart) [inline tests/integration/main.go:140]: Shouldn't we spawn a goroutine here or at the call site? Depends exactly how you're intending to design your concurrency tho

File: tests/integration/main.go [hunk 6]
```diff
@@ -24,45 +25,47 @@ var localSdkPath = flag.String("local-sdk-path", "",
...
+		tr.teardownDocker()
+		return
+	}
 
 	start := time.Now()
 	tr := DefaultTestRun()
 	tr.SetLocalSDKPath(*localSdkPath)
 	tr.ValidateStringLiterals()
 	tr.startDocker()
+	tr.ExecuteSteps(happyPathSteps)
```
Comment (smarshall-spitzbart) [inline tests/integration/main.go:43]: Could we add a comment or TODO that currently everything runs sequentially, but we'll build out concurrency eventually

File: tests/integration/main.go [hunk 7]
```diff
@@ -131,9 +134,14 @@ func (tr *TestRun) runStep(step Step, verbose bool) {
 	}
 }
 
-// Starts docker container and sequentially runs steps
-func (tr *TestRun) ExecuteSteps(wg *sync.WaitGroup, steps []Step) {
+// ExecuteStepsInWaitGroup runs steps in a WaitGroup so that the test run can be run in parallel.
+func (tr *TestRun) ExecuteStepsInWaitGroup(wg *sync.WaitGroup, steps []Step) {
```
Comment (smarshall-spitzbart) [inline tests/integration/main.go:138]: Seems less bug prone if you don't require the caller to increment the wg themselves
```suggestion
func (tr *TestRun) ExecuteStepsInWaitGroup(wg *sync.WaitGroup, steps []Step) {
wg.Add(1)
```

File: tests/integration/main.go [hunk 8]
```diff
@@ -184,3 +192,16 @@ func (tr *TestRun) startDocker() {
 	err = cmd.Wait()
 	log.Fatalf("StartDocker exited with error: %v", err)
 }
+
+// remove docker container to reduce resource usage
+// otherwise the chain will keep running in the background
+func (tr *TestRun) teardownDocker() {
```
Comment (smarshall-spitzbart) [inline tests/integration/main.go:198]: Nice 👍

**Review by smarshall-spitzbart — COMMENTED** [review]
Some comments to address but overall looks solid, thanks!

fyi I'm not able to run either version locally
```
shawn@Shawns-MacBook-Pro interchain-security % make test-integration-full
go run ./tests/integration/... --multiconsumer
=============== building default testRun ===============
2023/01/11 12:29:36 StartDocker exited with error: exit status 1
exit status 1
make: *** [test-integration-full] Error 1
```

```
shawn@Shawns-MacBook-Pro interchain-security % make test-integration
go run ./tests/integration/... --happy-path-only
=============== building default testRun ===============
2023/01/11 12:30:33 StartDocker exited with error: exit status 1
exit status 1
make: *** [test-integration] Error 1
```

**Comment by smarshall-spitzbart** [comment]
If running integration tests locally is not a problem for anyone else, feel free to move on and I'll fix my docker issues after vacay

**Comment by MSalopek** [comment]
> Some comments to address but overall looks solid, thanks!
> 
> fyi I'm not able to run either version locally
> 
> ```
> shawn@Shawns-MacBook-Pro interchain-security % make test-integration-full
> go run ./tests/integration/... --multiconsumer
> =============== building default testRun ===============
> 2023/01/11 12:29:36 StartDocker exited with error: exit status 1
> exit status 1
> make: *** [test-integration-full] Error 1
> ```
> 
> ```
> shawn@Shawns-MacBook-Pro interchain-security % make test-integration
> go run ./tests/integration/... --happy-path-only
> =============== building default testRun ===============
> 2023/01/11 12:30:33 StartDocker exited with error: exit status 1
> exit status 1
> make: *** [test-integration] Error 1
> ```

Maybe try pruning your local containers and running `go run ./tests/integration --verbose`.

Verbose will describe each build step

File: tests/integration/main.go (hunk 5 above)
Comment (MSalopek) [reply tests/integration/main.go]: Added option to run tests in parallel or in sequence using `--parallel` flag

File: tests/integration/main.go (hunk 7 above)
Comment (MSalopek) [reply tests/integration/main.go]: `ExecuteStepsInWaitGroup` was removed.

Both `wg.Add(1)` and `wg.Done()` are called from the top level after refactoring.
```golang
wg.Add(1)
go func(run testRunWithSteps) {
	defer wg.Done()
	tr := run.testRun
	tr.SetLocalSDKPath(*localSdkPath)
	tr.ValidateStringLiterals()
	tr.startDocker()
	tr.ExecuteSteps(run.steps)
	tr.teardownDocker()
}(run)
```

File: tests/integration/main.go [hunk 9]
```diff
@@ -24,48 +26,63 @@ var localSdkPath = flag.String("local-sdk-path", "",
...
+	start := time.Now()
+	if parallel != nil && *parallel {
+		fmt.Println("=============== running all tests in parallel ===============")
+		var wg sync.WaitGroup
+		for _, run := range testRuns {
+			wg.Add(1)
+			go func(run testRunWithSteps) {
+				defer wg.Done()
+				tr := run.testRun
+				tr.SetLocalSDKPath(*localSdkPath)
```
Comment (glnro) [inline tests/integration/main.go:58]: Could probably abstract L:58-L:62 out as it's in all three run scopes

File: tests/integration/main.go [hunk 10]
```diff
@@ -24,46 +26,63 @@ var localSdkPath = flag.String("local-sdk-path", "",
...
-	if *multiconsumer {
-		mul := MultiConsumerTestRun()
-		mul.SetLocalSDKPath(*localSdkPath)
-		mul.ValidateStringLiterals()
-		mul.startDocker()
+	if happyPathOnly != nil && *happyPathOnly {
+		fmt.Println("=============== running happy path only ===============")
+		tr := DefaultTestRun()
+		tr.SetLocalSDKPath(*localSdkPath)
+		tr.ValidateStringLiterals()
```
Comment (sainoe) [inline tests/integration/main.go:33]: These four lines might be reformatted to:
```
tr.Run(happyPathSteps, *localSdkPath)
```

File: tests/integration/main.go [hunk 11]
```diff
@@ -24,46 +26,63 @@ var localSdkPath = flag.String("local-sdk-path", "",
...
+	for _, run := range testRuns {
+		tr := run.testRun
+		tr.Run(run.steps, *localSdkPath)
+	}
+	fmt.Printf("TOTAL TIME ELAPSED: %v\n", time.Since(start))
+}
 
-	wg.Add(1)
-	go slash.ExecuteSteps(&wg, slashThrottleSteps)
+// Run sets up docker container and executs the steps in the test run.
```
Comment (sainoe) [inline tests/integration/main.go:73]:
```suggestion
// Run sets up docker container and executes the steps in the test run.
```

File: tests/integration/main.go [hunk 12]
```diff
@@ -24,46 +26,63 @@ var localSdkPath = flag.String("local-sdk-path", "",
...
+// Docker containers are torn down after the test run is complete.
+func (tr *TestRun) Run(steps []Step, localSdkPath string) {
```
Comment (sainoe) [inline tests/integration/main.go:75]: nit: it would be nice to have all  the`TestRun` methods unexported for consistency.

**Review by sainoe — APPROVED** [review]
LGTM, just few suggestions. Nice work @MSalopek.

File: tests/integration/main.go (hunk 10 above)
Comment (MSalopek) [reply tests/integration/main.go]: Added

File: tests/integration/main.go (hunk 11 above)
Comment (MSalopek) [reply tests/integration/main.go]: Done

File: tests/integration/main.go (hunk 12 above)
Comment (MSalopek) [reply tests/integration/main.go:71]: Done

File: tests/integration/main.go [hunk 13]
```diff
@@ -131,9 +146,8 @@ func (tr *TestRun) runStep(step Step, verbose bool) {
 	}
 }
```
Comment (yaruwangway) [inline tests/integration/main.go:148]: line 133, 
log.Fatalf(fmt.Sprintf(`unknown action in testRun %s: %#v`, tr.name, action)) -> log.Fatalf(`unknown action in testRun %s: %#v`, tr.name, action) ?

Comment (MSalopek) [reply tests/integration/main.go:148]: Changed :)

**Review by glnro — APPROVED** [review]

**Pull request merged**